app/lib/supabase/supabase-realtime.ts

```diff
@@ -0,0 +1,113 @@
+import {
+  REALTIME_SUBSCRIBE_STATES,
+  type RealtimeChannel,
+} from '@supabase/supabase-js';
+import { use } from 'react';
+import { type StoreApi, type UseBoundStore, create } from 'zustand';
+import { useEffectNoStrictMode } from '~/hooks/use-effect-no-strict-mode';
+
+type SupabaseRealtimeSubscriptionStore = {
+  channel: RealtimeChannel | null;
+  subscriptionPromise: Promise<RealtimeChannel> | null;
+  subscriptionError: Error | null;
+  promiseSettled: boolean;
+  subscribe: (
+    channelFactory: () => RealtimeChannel,
+  ) => Promise<RealtimeChannel>;
+  unsubscribe: () => void;
+};
+
+export const createSupabaseRealtimeSubscriptionStore = () =>
+  create<SupabaseRealtimeSubscriptionStore>((set, get) => ({
+    channel: null,
+    subscriptionPromise: null,
+    promiseSettled: false,
+    subscriptionError: null,
+    subscribe: (
+      channelFactory: () => RealtimeChannel,
+    ): Promise<RealtimeChannel> => {
+      let subscriptionPromise = get().subscriptionPromise;
+      if (subscriptionPromise) {
+        return subscriptionPromise;
+      }
+
+      subscriptionPromise = new Promise((resolve, reject) => {
+        const channel = channelFactory().subscribe((status, error) => {
+          if (!get().promiseSettled) {
```

Comment:
if promise is already settled but this callback is called again that will be for any changes to subscription status that happen after initial "handshake". e.g. when there is some error that breaks the subscription etc. In that case what I did is I set the `subscriptionError` in the store and then hook below will detect that error and throw it which will then be caught by our global error boundary